src/pybind/mgr/dashboard/services/nvmeof_client.py

```diff
@@ -0,0 +1,140 @@
+from enum import Enum
+from typing import Optional
+import grpc
+import json
+
+import logging
+
+from .proto import gateway_pb2 as pb2
+from .proto import gateway_pb2_grpc as pb2_grpc
+
+from google.protobuf.json_format import MessageToJson
+
+from .nvmeof_conf import NvmeofGatewaysConfig
+from ..tools import str_to_bool
+
+logger = logging.getLogger('nvmeof_client')
+
+
+class NVMeoFClient(object):
+    def __init__(self):
+        logger.info('Initiating nvmeof gateway connection...')
+
+        self.gateway_addr = list(NvmeofGatewaysConfig.get_gateways_config()['gateways'].values())[0]['service_url']
+        self.channel = grpc.insecure_channel(
+            '{}'.format(self.gateway_addr)
+        )
+        logger.info('Found nvmeof gateway at {}'.format(self.gateway_addr))
+        self.stub = pb2_grpc.GatewayStub(self.channel)
+
+    def list_subsystems(self, subsystem_nqn: Optional[str] = None, serial_number: Optional[str] = None):
+        if subsystem_nqn is not None and serial_number is not None:
+            response = self.stub.list_subsystems(pb2.list_subsystems_req(
+                subsystem_nqn=subsystem_nqn,
+                serial_number=serial_number
+            ))
+        response = self.stub.list_subsystems(pb2.list_subsystems_req())
+        response = json.loads(MessageToJson(response))
```

Comment:
Why this ` json.loads(MessageToJson())`? So we are converting:
Protobuf Message -> JSON string -> Python Dict -> JSON string (the object returned by all Dashboard controllers is serialized to JSON).

Isn't there a way to just pick the `subsystems` field without further json/dict deserialization? Doesn't `return reponse.subsystems` simply work?

Reply:
An alternative approach would be to use the `MessageToDict()` but Python `dicts` have probably more overhead than plain JSON objects, and they'll be eventually converted to JSON. However, if we return a JSON object by simply doing the following, we'll get a doubly-JSON encoded object:
```suggestion
        response = MessageToJson(response.response)
```

Something like:
```py
>>> import json

>>> o = dict(foo="bar")

>>> json.dumps(o)
'{"foo": "bar"}'

>>> json.dumps(json.dumps(o))
'"{\\"foo\\": \\"bar\\"}"'
```

So the alternative would be to either:
* Create a decorator similar to the `@allow_empty_body` (that's for request JSON object, since the request/input is automatically converted by Cherrypy's `json_in` Tool), but for the response (output), to stop `__request_wrapper()` at `_base_controller.py` from using `json.dumps()` for objects that are already valid JSON. It might be something like `@skip_response_encoding` or something like that.
* Or change that `__request_wrapper()` to detect: if the `ret` object is a valid JSON object (e.g.: is a `str` AND `json.loads()`). This would be more automatic, but on the other hand, it results in parsing a whole JSON string and creating a Python object from it...

Wrapping up (btw, the JSON encoding should be in the controller, not in the service, since it's controllers who care about input/output encodings (services should just provide valid Python objects):
```py
@APIRouter('/nvmeof/subsystem', Scope.NVME_OF)
@APIDoc('NVMe-oF Subsystem Management API', 'NVMe-oF')
class NvmeofSubsystem(RESTController):
    @skip_response_encoding
    @ReadPermission
    @EndpointDoc("List all NVMeoF gateways",
                 parameters={
                     'subsystem_nqn': (str, 'NVMeoF subsystem NQN'),
                     'serial_number': (str, 'NVMeoF subsystem serial number')
                 })
    def list(self, subsystem_nqn: Optional[str]= None, serial_number: Optional[str] = None):
        return MessageToJson(NVMeoFClient().list_subsystems(
            subsystem_nqn=subsystem_nqn, serial_number=serial_number))
```

What do you think?